core/quantity/src/Parser.ts

```diff
@@ -3,21 +3,49 @@
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
 import { QuantityConstants } from "./Constants";
-import { QuantityStatus } from "./Exception";
 import { Format } from "./Formatter/Format";
 import { FormatTraits, FormatType } from "./Formatter/FormatEnums";
 import { PotentialParseUnit, QuantityProps, UnitConversion, UnitConversionSpec, UnitProps, UnitsProvider } from "./Interfaces";
 import { ParserSpec } from "./ParserSpec";
 import { Quantity } from "./Quantity";
 
+/** Possible parser errors
+ * @alpha
+ */
+export enum ParseError {
+  UnableToGenerateParseTokens = 1,
+  NoValueOrUnitFoundInString,
+  UnitLabelSuppliedButNotMatched,
+  UnknownUnit,
+  UnableToConvertParseTokensToQuantity,
+  InvalidParserSpec,
+}
+
+/** Parse error result from [[Parser.parseToQuantityValue]] or [[Parser.parseToQuantityValue]].
+ * @alpha
+ */
+export interface ParseQuantityError {
+  /** Union discriminator for [[QuantityParseResult]]. */
+  ok: false;
+  /** The specific error that occurred during parsing. */
+  error: ParseError;
+}
+
+/** Successful result from [[Parser.parseToQuantityValue]] or [[Parser.parseToQuantityValue]].
+ * @alpha
+ */
+export interface ParsedQuantity {
+  /** Union discriminator for [[QuantityParseResult]]. */
+  ok: true;
+  /** The magnitude of the parsed quantity. */
+  value: number;
+}
+
 /**
  * Defines Results of parsing a string input by a user into its desired value type
  * @alpha
  */
-export interface ParseResult {
-  value?: number | undefined;
-  status: QuantityStatus;
-}
+export type QuantityParseResult = ParsedQuantity | ParseQuantityError;

```

Comment:
I like the new use of the discriminated union but it feels a bit weird to have the methods on `Parser` to figure out if the returned value is actually the Quantity or an error.

What do you think of switching to something like below?  Not sure if I actually like this approach better since it requires a class (more overhead) but it does allow you to only need one object and consolidate the API.

Instead of the above 3 types (`QuantityParseResult`, `ParsedQuantity`  and `ParseQuantityError`) you would have something like,

```ts
export type QuantityParseResult = ParsedQuantityResult<ParseError, number> | ParsedQuantityError<ParseError, number>;

class ParsedQuantityResult<E, V> {
  readonly value: V;

  constructor(value: V) {
    this.value = value;
  }

  isQuantity(): this is ParsedQuantityResult<E, V> {
    return true;
  }

  isError(): this is ParsedQuantityError<E, V> {
    return false;
  }
}

class ParsedQuantityError<E, V> {
  readonly error: E;

  constructor(error: E) {
    this.error = error;
  }

  isQuantity(): this is ParsedQuantityResult<E, V> {
    return false;
  }

  isError(): this is ParsedQuantityError<E, V> {
    return true;
  }
}
```

This means that the resulting `QuantityParseResult` could be used in an Either scenario by the distinction of the two readonly properties.  Which allows the following type of behavior from an API consumer perspective:

```ts
const parseResult = Parser.parseQuantityString(testEntry.value, parserSpec);
assert.isTrue(parseResult.isQuantity());
if (parseResult.isQuantity()) {
  assert.isTrue(Math.abs(parseResult.value - testEntry.magnitude) < 0.0001);
}
``` 

This should work in a similar way to what you have now but instead of calling `Parser.isQuantity` you can just use the result object to check.

Reply:
@bsteinbk and I discussed offline and since this requires a class to be constructed, we'll skip doing it for now until we have more of a reason to add the overhead.